src/mlpack/tests/main_tests/fastmks_test.cpp

```diff
@@ -0,0 +1,657 @@
+}
+
+/*
+ * Ensure that single-tree search returns the same result
+ * as dual-tree search
+ */
+BOOST_AUTO_TEST_CASE(FastMKSTreeTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("k", (int) 10);
+
+  mlpackMain();
+
+  arma::Mat<size_t> indices;
+  arma::mat kernel;
+  indices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  kernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  bindings::tests::CleanMemory();
+
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["k"].wasPassed = false;
+
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("single", true);
+
+  mlpackMain();
+
+  CheckMatrices(indices,
+      CLI::GetParam<arma::Mat<size_t>>("indices"));
+  CheckMatrices(kernel,
+      CLI::GetParam<arma::mat>("kernels"));
+}
+
+/*
+ * Ensure that we get almost same results in cover tree
+ * search mode when different basis is specified.
+ */
+BOOST_AUTO_TEST_CASE(FastMKSBasisTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, some k <= number of reference points.
+  SetInputParam("reference", referenceData);
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 3.0);
+
+  mlpackMain();
+
+  arma::Mat<size_t> indices;
+  arma::mat kernel;
+  indices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  kernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  bindings::tests::CleanMemory();
+
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["k"].wasPassed = false;
+
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 4.0);
+
+  mlpackMain();
+
+  arma::Mat<size_t> newindices;
+  arma::mat newkernel;
+  newindices = std::move(CLI::GetParam<arma::Mat<size_t>>("indices"));
+  newkernel = std::move(CLI::GetParam<arma::mat>("kernels"));
+
+  CheckMatrices(indices, newindices);
+  CheckMatrices(kernel, newkernel);
+}
+
+/**
+  * Check that we can't specify base less than 1.
+  */
+BOOST_AUTO_TEST_CASE(FastMKSBaseTest)
+{
+  // 100 points in 3 dimensions.
+  arma::mat referenceData(3, 100, arma::fill::randu);
+
+  // Random input, invalid base.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 10);
+  SetInputParam("base", 0.0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::invalid_argument);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+  * Ensure that different kernels returns different results.
+  */
+BOOST_AUTO_TEST_CASE(FastMKSKernelTest)
+{
+  std::string kerneltypes[] = {"polynomial", "cosine", "gaussian",
+      "epanechnikov", "triangular", "hyptan"};
```

Comment:
If this is the test that is running forever for `hyptan`, I'd suggest a different strategy for `hyptan`.  I think you should be able to use the `vc2` dataset and that is more likely to work:

```
  arma::mat inputData;
  if (!data::Load("vc2.csv", inputData))
    BOOST_FAIL("Cannot load test dataset vc2.csv!");
```

Then you can use the `vc2` dataset for both the reference and query set.